Automattic/minifiers
# Run swc.minifySync inside a pool of forked child processes

### PR Dependency Note

* This includes the prettier changes as well. If #5 is merged, this PR will be shorter and easier to read.

### Summary

**Fork child processes when doing JS minification. This contains the memory leak.**

* We use a pool of child processes and use the `generic-pool` library to help manage this.
* New files are lin `lib/processes`.

There is also some new optional debugging to help see what's going on, opted in via environment flags.

### Test Plan

- Turn on the new debuggers with this in bash:

```
export DEBUG_MEMORY=1
export DEBUG_POOL=1
export DEBUG_QUIET_REQUEST=1
```

- Run the server
  - `npm start`
- Clone the load tester: https://github.com/mreishus/test-mini-js 
  - In a 2nd terminal window, cd into `files` and run `./serve.sh` in test-mini-js
  - In a 3rd terminal window, run `node main.js` in test-mini-js. This should send lots of requests to the server
- Double check the minifier is working in your browser: Visit `http://127.0.0.1:4747/get?url=http%3A%2F%2F127.0.0.1%3A1234%2Freact.production.min.js&random=123asdf`
- Watch the memory usage. It should not be unbounded.

To compare to the old versions, shut off the server, make this change:
```diff
diff --git a/config.js b/config.js
index c7c9a7c..6458499 100644
--- a/config.js
+++ b/config.js
@@ -14,8 +14,8 @@ module.exports = () => {
                        regex: new RegExp(
                                /^(application\/javascript)|(application\/x-javascript)|(application\/ecmascript)|(text\/javascript)|(text\/ecmascript)/
                        ),
-                       min_func: require( './lib/processes/js-min-parent.js' ), // Forking version
-                       // min_func: require( './lib/js-min.js' ), // Non-forking version
+                       //min_func: require( './lib/processes/js-min-parent.js' ), // Forking version
+                       min_func: require( './lib/js-min.js' ), // Non-forking version
                        type: 'string',
                },
                json: {
```
Re-do the steps, and you should see the memory usage become unbounded.
